src/mirage_vm.ml

```diff
@@ -27,33 +27,61 @@ module Main (C: V1_LWT.CONSOLE) = struct
    * formatting instructions.
   *)
   let log c fmt = Printf.kprintf (fun msg -> C.log_s c msg) fmt
-
+  let ack' client path  = XS.(immediate client @@ fun h -> write h path "") 
   let read client path = XS.(immediate client @@ fun h -> read h path) 
-  let ack  client path = XS.(immediate client @@ fun h -> write h path "") 
+
+  (* [ack] acknowledges a message and offers to violate the proper
+   * protocol (AckOK) by doing something else *)
+
+  let ack client path = function
+    | CMD.AckOK        -> XS.(immediate client @@ fun h -> write h path "") 
+    | CMD.AckWrite(x)  -> XS.(immediate client @@ fun h -> write h path x )
+    | CMD.AckNone      -> return () (* do nothing *)
+    | CMD.AckDelete    -> XS.(immediate client @@ fun h -> rm h path)
+
 
   (* [read_opt client path] reads [path] from the Xen Store and
    * returns it as an option value on success, and [None] otherwise.
-   * Unexpected errors still raise an exception.
+   * A empty string is returned as [None] (and thus conflates
+   * no string and the empty string).  Unexpected errors still raise an
+   * exception.
   *)
   let read_opt client  path  = 
     Lwt.catch
       ( fun () ->
-          read client path >>= fun msg -> 
-          return (Some msg)
+          read client path >>= 
+          ( function 
+          | ""  -> return None       (* XXX right design choice? *)
```

Comment:
I think you are right to be concerned about conflating the empty string and ENOENT. If you returned `Some ""` it could still be pattern-matched.